optimum/onnxruntime/base.py

```diff
@@ -11,453 +11,592 @@
-
-    _prepare_onnx_inputs = ORTModel._prepare_onnx_inputs
-    _prepare_onnx_outputs = ORTModel._prepare_onnx_outputs
+    def initialize_ort_attributes(self, session: InferenceSession, use_io_binding: Optional[bool] = None):
+        """
+        Initializes the ORTSessionMixin class.
+        Args:
+            session (`onnxruntime.InferenceSession`):
+                The ONNX Runtime session to use for inference.
+            use_io_binding (`Optional[bool]`, defaults to `None`):
+                Whether to use IO Binding or not. If `None`, it will be set to `True` for CUDAExecutionProvider and `False`
+                for other providers.
+        """
 
-    def __init__(self, session: InferenceSession, parent_model: "ORTModel"):
         self.session = session
-        self.parent_model = parent_model
-        self.main_input_name = self.parent_model.main_input_name
 
-        self.input_names = {input_key.name: idx for idx, input_key in enumerate(self.session.get_inputs())}
-        self.output_names = {output_key.name: idx for idx, output_key in enumerate(self.session.get_outputs())}
+        if use_io_binding is None:
+            if self.provider == "CUDAExecutionProvider":
+                logger.info(
+                    "`use_io_binding` was not set, but CUDAExecutionProvider supports IO Binding. "
+                    "Setting `use_io_binding=True` to leverage IO Binding and improve performance. "
+                    "You can disable it by setting `model.use_io_binding=False`."
+                )
+                use_io_binding = True
+            else:
+                use_io_binding = False
 
-        self.input_dtypes = {input_key.name: input_key.type for input_key in session.get_inputs()}
-        self.output_dtypes = {output_key.name: output_key.type for output_key in session.get_outputs()}
+        self._use_io_binding = use_io_binding
+        self._io_binding = IOBinding(session)
+        self._dtype = get_dtype_from_session(session)
+        self._device = get_device_for_provider(self.provider, self.provider_option)
 
-        self.input_shapes = {input_key.name: input_key.shape for input_key in session.get_inputs()}
-        self.output_shapes = {output_key.name: output_key.shape for output_key in session.get_outputs()}
+        self.input_names = {input.name: idx for idx, input in enumerate(session.get_inputs())}
+        self.output_names = {output.name: idx for idx, output in enumerate(session.get_outputs())}
+        self.input_shapes = {input.name: input.shape for input in session.get_inputs()}
+        self.output_shapes = {output.name: output.shape for output in session.get_outputs()}
+        self.input_dtypes = {input.name: input.type for input in session.get_inputs()}
+        self.output_dtypes = {output.name: output.type for output in session.get_outputs()}
+
+        self.model_path = Path(session._model_path)
+        self.model_name = self.model_path.name
```

Comment:
doesn't seem to be used anywhere so we could potentially remove it in a future release (adding a warning in the meantime)

Reply:
yes talking about `model_name` which seems to be used in the tests (and can be easily extracted from `model_path`) so we can remove imo, not specific to this PR though so feel free to close!